src/circuit_proof/mod.rs

```diff
@@ -0,0 +1,781 @@
+#![allow(non_snake_case)]
+
+use rand::{CryptoRng, Rng};
+use std::iter;
+
+use curve25519_dalek::ristretto::RistrettoPoint;
+use curve25519_dalek::scalar::Scalar;
+use curve25519_dalek::traits::{IsIdentity, MultiscalarMul, VartimeMultiscalarMul};
+
+use generators::Generators;
+use inner_product_proof::{inner_product, InnerProductProof};
+use proof_transcript::ProofTranscript;
+use util;
+
+#[derive(Clone, Debug)]
+pub struct Circuit {
+    n: usize,
+    m: usize,
+    q: usize,
+    c: Vec<Scalar>,        // vector of constants, length q
+    W_L: Vec<Vec<Scalar>>, // q vectors, of length n each
+    W_R: Vec<Vec<Scalar>>,
+    W_O: Vec<Vec<Scalar>>,
+    W_V: Vec<Vec<Scalar>>, // q vectors, of length m each
```

Comment:
I think it would be good to put a note that we should swap out this representation later when we have R1CS support, since for large circuits we almost certainly don't want to actually keep all of these matrices in memory.